Spot Light Tool

ericmehl opened the pull request:
This adds a tool in the scene viewer for manipulating spot light cone and penumbra angles.

There are a few pieces missing :
- UI at the bottom of the viewer for showing what plug(s) you are manipulating, edit warnings, etc. as the transform tools do.
- A better graphic for the handles so they are easier to tell apart from each other.
- Better handling of dragging the penumbra handle when looking through a light with either `inset` or `outset` modes.

But the code that's there now is ready to be reviewed. If it's easier to take care of those todos above in a separate PR so this doesn't grow to big, I'm happy doing that.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1266 @@
+
+	public :
+
+		/// \todo SpotLightTool (below) is not using `metadataKey` because it knows what keys it needs.
+		/// If that is going to be the case for other tools, we should remove it.
```
Keeping the base class as pure as possible seems sensible to me, with as few assumptions about what a derived class might want to do as possible. Even `attributePattern` seems a bit specific - my original assumption was that an Inspector was the key thing the base class would require, so that we've abstracted away what is being queried.

johnhaddon commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1266 @@
+	return solidArc;
+}
+
+class LightToolHandle : public Handle
```
It would be great to document the basic design intent of LightToolHandle here, with some discussion of how and why responsibilities are divvied up between LightToolHandle and LightTool. It's probably my failing, but I haven't quite got the "big picture" yet.

One thing I'm wondering about is why LightToolThingy _is a_ Handle, rather than why LightToolThingy _has a_ Handle. The original Handle design assumed that the handle provided drag information for a separate client, with the intention that handles could be reused by different clients.

In case it's not clear, I'm not advocating for a redesign at this point, just trying to understand the current design and the reasons behind it.

johnhaddon commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1266 @@
+			return m_lookThroughLight;
+		}
+
+		// Called at the beginning of a drag with the path to be inspected
+		// set in the current context. Must be implemented by derived classes to collect
+		// whatever data will be needed to implement the drag operation.
+		virtual void addDragInspection() = 0;
+
+		virtual bool handleDragMove( const GafferUI::DragDropEvent &event ) = 0;
+		virtual bool handleDragEnd() = 0;
```
What's the reason for these being public methods called by LightTool, rather than protected methods called by LightToolHandle?

johnhaddon commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1266 @@
+}
+
+void LightTool::updateSelection() const
+{
+	if( !m_selectionDirty )
+	{
+		return;
+	}
+
+	if( m_dragging )
+	{
+		// In theory, an expression or some such could change the effective
+		// transform plug while we're dragging (for instance, by driving the
+		// enabled status of a downstream transform using the translate value
+		// we're editing). But we ignore that on the grounds that it's unlikely,
+		// and also that it would be very confusing for the selection to be
+		// changed mid-drag.
+		return;
+	}
+
+	m_selection.clear();
+	m_selectionDirty = false;
+
+	if( !activePlug()->getValue() )
+	{
+		return;
+	}
+
+	// If there's no input scene, then there's no need to
+	// do anything. Our `scenePlug()` receives its input
+	// from the View's input, but that doesn't count.
+	auto scene = scenePlug()->getInput<ScenePlug>();
+	scene = scene ? scene->getInput<ScenePlug>() : scene;
+	if( !scene )
+	{
+		return;
+	}
+
+	m_selection = ContextAlgo::getSelectedPaths( view()->getContext() );
+}
```
This has its origins in TransformTool, where updating the selection had significant overhead (because really it was doing inspections). Since all we do here is retrieve a pointer from the context, I'm not sure what `m_selectionDirty` is achieving - perhaps this could be simpler?

johnhaddon reviewed (commented):
Thanks Eric!

I haven't done a very detailed review, partly because I'm a bit fuzzy headed today, and also because I'd like to learn more about the high-level design before looking more closely at the details. So most of my comments are just asking questions to help me understand better rather than making any concrete suggestions - sorry if this should all have been obvious to me.

@murraystevenson also had a few comments after testing interactively (mostly about the conundrum of where the handles belong on the cone/frustum). I think it'd be good to review those with him before jumping back into the code, just in case they inform any changes to the design...

Cheers...
John

ericmehl commented:
Noting here that the commit history of this PR has some commits from #5281 which can be dropped when the interconnections between these two PRs are resolved and that one is merged.

ericmehl commented:
Ready for review!

The relevant commits are the change to `Inspector` to consider read only plugs, done in c198c1132ecfee755bec93a59fa0925d2a8d6ba6, and the `LightTool` commit itself.

This will also depend on what settles out for the change to plug dirty handling currently done in https://github.com/GafferHQ/gaffer/pull/5218/commits/53057d77216bc71e9024fb8e444f76538d363690.

danieldresser-ie commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1841 @@
+			// against divide-by-zero. Penumbra handles are hidden close to zero, so they
+			// don't need to be checked.
+
+			for( auto &[coneInspection, originalConeAngle, penumbraInspection, originalPenumbraAngle] : m_inspections )
```
I really don't much like this loop. It seems the logic it's implementing is legitimately quite complicated, and some messiness is probably required here, but the current structure leaves me with a bunch of questions:
* is angleScale ever < 0? Looking through all the cases, it seems like the times when we modify the second value in max, it is to raise it above 0? If that's true, then we could put one max with 0 at the end, and remove it from inside most branches.
* is there ever a situation where the second value that is passed to min on one loop iteration is less than the second value passed to max on another iteration? Then the initial value of angleScale would be ignored, and the resulting value would depend solely on which order m_inspections are visited in, which seems bad. It's very likely that this can't happen - but I don't see any easy way to prove it can't by quickly inspecting this code.
* It may be a bit late for any reconsideration of user behaviour, if this has already been validated, but I do wonder whether this is the simplest and most desirable approach. First off, to confirm I'm understanding what this is currently intended to do in the simplest case: suppose I have light A with a cone angle of 60, and light B with a cone angle of 120, and I'm using a penumbra handle with absolute mode, and both penumbra's are currently set the same as the cone angle. Because light B has a large cone angle, the maximum angleScale permitted is `180/120 = 1.5`. This means that if editing light A at the same time, if I drag A's penumbra larger, it will be locked to a maximum of `60 * 1.5 = 90` degrees because of the limitation on B, right? Is this actually desirable? The more natural approach to me would seem to be to not clamp angleScale, but for each selected target, apply angleScale, then apply appropriate clamping ... it seems like the logic would be a lot simpler if just clamping one at a time, and I'm not sure what the advantage is of clamping out light A because light B has hit its maximum.

ericmehl replied:
I'm not 100% convinced this is a better behavior, but for consideration it's in https://github.com/GafferHQ/gaffer/pull/5218/commits/a8fb0e07f5e018555f665db99bd48a9be81cdff9.

My hesitation is in the case of :
1. Two lights in the same scene, one with a penumbra value that is pretty close to the cone angle (so not far for it to go before it's clamped).
2. Drag the cone handle whose penumbra is almost clamped, it will move a few degrees and then stop as expected.
3. The other light is still moving because it hasn't been clamped, but there's no feedback in the UI - the light with the handles is not changing, and neither is the arc indicator for the angles.

I'm not dead-set against independent clamping, especially with the elimination of the scaling behavior of the values this allows, but that's my main point against it at the moment.

danieldresser-ie replied:
Hmm, that's a fair concern. The code does seem a whole lot clearer this way, but it looks like maybe part of that is a refactor that could be kept even if the behaviour went back to the old behaviour?

I suppose what your point suggests is clamping based on the light with the handles, and the light being edited, but maybe not based on all selected lights? i.e. in your situation 2, both lights would clamp after the handle reaches maximum, but if you dragged the handle on the other light instead, then you could keep dragging larger even after the almost-clamped penumbra has started clamping?

For any of these approaches, maybe we need to to package up clamping into a function that can perform clamping for an m_inspection, and can be called wherever we decide it's needed?

ericmehl replied:
I like the idea of clamping based on either the active light or each light individually, whichever comes first. I implemented that in https://github.com/GafferHQ/gaffer/pull/5218/commits/416630b7210cf40226374f9fe9716279a47a84b8.

It was a bit of a rabbit hole because I wasn't holding onto _both_ the cone and penumbra data for the active light, both of which are needed to figure out the clamping. But I think my solution also improved the general data handling, and forced me to be more explicit about when cone angles are for the plug and for the handle (the difference between what the renderer expects and what the handle represents).

Settling on manipulating handle-space values until the last minute setting of a plug allowed me to remove all of the `handleAngle` calls too.

Finally, I extracted the clamp function to its own method so it's reusable.

All of that is in the final commit, https://github.com/GafferHQ/gaffer/pull/5218/commits/416630b7210cf40226374f9fe9716279a47a84b8, passing my use-tests and I think feeling pretty good!

danieldresser-ie replied:
This all looks quite good to me, and I think the behaviour makes a lot of sense, hopefully John agrees.

There is one minor stylistic thing, which is minor enough that I wouldn't recommend changing it here, but for future reference: in this last commit, you've switched originalPenumbraAngleData ( which is just used as an input to handleAngle() ) from float to FloatDataPtr - and it appears you don't need any of the runtime polymorphism of Data, it's just to have an optional value? I think I would prefer std::optional here ( the performance isn't important, but it documents intent more clearly ), especially since the output from handleAngle() is an std::optional - but it's a slight preference.

johnhaddon replied:
> This all looks quite good to me, and I think the behaviour makes a lot of sense, hopefully John agrees.

The latest clamping behaviour you both described sounds good to me.

danieldresser-ie reviewed (commented):
That's a lot of code! Tried to read through most of it, focusing on math stuff.

Didn't spot anything wrong, but there were a few places I've asked about clarification.

ericmehl commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1841 @@
+		""
+		"	vec3 pAimed = vertexP.x * aimedXAxis + vertexP.y * aimedYAxis + vertexP.z * aimedZAxis;"
+		""
+		"	vec4 pCam = gl_ModelViewMatrix * vec4( pAimed, 1 );"
```
Yes, the intention is to remove the effects of scale in the `gl_ModelViewMatrix`. This vertex shader is used by the handle widgets, and we want them to draw the same size on screen independent of the view and handle transformations.

ericmehl commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1841 @@
+
+			const V2f gadgetRasterOrigin = m_view->viewportGadget()->gadgetToRasterSpace( V3f( 0, 0, -1.f ), this );
+			const V2f rasterSphereIntersection = m_view->viewportGadget()->gadgetToRasterSpace( sphereIntersection, this );
+			const V2f rasterNormal = ( m_view->viewportGadget()->gadgetToRasterSpace( V3f( 0, 1.f, -1.f ), this ) - gadgetRasterOrigin ).normalized();
```
It would be more intuitive if we could do `gadgetToRasterSpace( V3f( 0, 1, 0 ) )` but `gadgetToRasterSpace` divides by the (negative) z coordinate for perspective projection. So using `-1` for Z here compensates for that transform.

Perhaps it needs a better name, or a comment explaining it?

danieldresser-ie replied:
Looking at this again ... I think it's fine. We're subtracting from gadgetRasterOrigin that also has the -1 Z value, that totally makes sense and I'm not sure why I missed this before.

ericmehl commented on include/GafferUI/Handle.h:
```diff
@@ -171,7 +171,7 @@ class GAFFERUI_API Handle : public Gadget
 
 		// Helper for performing drags around a rotation axis.
-		struct AngularDrag
+		struct GAFFERUI_API AngularDrag
```
I pulled that out to its own commit, now in https://github.com/GafferHQ/gaffer/pull/5218/commits/1ce9f8cd45350db39064de45b3fb468c82a5af91

ericmehl commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1841 @@
+				// Intersect the gadget-local `event` line with the sphere centered at the gadget
+				// origin with radius equal to the distance along the handle where the user clicked.
+				// `Imath::Sphere3::intersect()` returns the closest (if any) intersection, but we
+				// want the intersection closest to the handle line, so we do the calculation here.
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5218/commits/b1fdcd92ebb8e6a6db99fd02d4465b45f2f57708

ericmehl commented:
I rebased and squashed the fixes down, which dropped the first few commits that were borrowed from #5281.

I think all that is left is the fix currently cherry picked in https://github.com/GafferHQ/gaffer/pull/5218/commits/f92d6064682e8cd564f71ee2f0d78a1ffa48a030 and this can be ready for a final look.

ericmehl commented:
I found one specific case that was not being handled correctly this afternoon and pushed what is hopefully 🤞 the last fixup.

The bug was in selecting multiple lights, some of which have their cone angle disabled. Dragging the handle, whether it's on an enabled or disabled light, would change the cone of those lights that _are_ enabled.

https://github.com/GafferHQ/gaffer/pull/5218/commits/4ce394cb1a245d34476dea870aabd9e1c3e6c759 returns early if all inspections are disabled, matching the behavior of the other transform tools.

murraystevenson commented:
Nice one Eric! I've taken another look at this from a usability perspective and there are two minor things that stand out but otherwise I think we're good to get this into people's hands for 1.3. :tada:

The light tool isn't scaling with the Viewer drawing overrides for Visualiser Scale and Light Frustum Scale. Local overrides made on a light are respected, but it's ignored when you set these globally from the Viewer toolbar.

![viewerFrustumScale](https://github.com/GafferHQ/gaffer/assets/50844517/09057ff6-71d7-412f-946d-153e28c296d4)

We also end up in a bit of an odd visual state when light tool is active on a light with its frustum turned off. We see the spokes drawn to the length of the frustum but with no circles from the end of the frustum to tie them together. I wonder if it's worth limiting the spoke length to that of the light cone when the frustum is disabled? (basically a frustum scale of 0.1)

![lightFrustumOff](https://github.com/GafferHQ/gaffer/assets/50844517/3961d1a5-093a-4314-975f-e87a1bbfe9cf)

ericmehl commented:
My latest fixup, https://github.com/GafferHQ/gaffer/pull/5218/commits/4001e68db2caa56551cb103248e72a8de538910e, handles the visualiser scale and light frustum scales. When we talked offline we settled on leaving the handles at their normal length even when the frustum is off on the belief that artists may still want to be able to use them. We can revisit if users start to find them intrusive.

Still passing my test suite and hopefully ready for a squash and merge!

murraystevenson commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1916 @@
+				penumbraAngleData = runTimeCast<const FloatData>( penumbraInspection->value() );
+			}
+
+			assert( coneAngleData );
```
This assert fails/a release build will crash in the following situation:

- Activate the light tool
- Select a spotLight
- Select any non-light object such as a mesh or camera while the light tool is active

```
gaffer: src/GafferSceneUI/LightTool.cpp:1324: {anonymous}::SpotLightHandle::DragStartData {anonymous}::SpotLightHandle::spotLightHandleAngles() const: Assertion `coneAngleData' failed.

__assert_fail () from /lib64/libc.so.6
(anonymous namespace)::SpotLightHandle::spotLightHandleAngles (this=0x7f50990be470) at src/GafferSceneUI/LightTool.cpp:1324
(anonymous namespace)::SpotLightHandle::visible (this=0x7f50990be470) at src/GafferSceneUI/LightTool.cpp:863
GafferSceneUI::LightTool::updateHandleInspections (this=0x7f5093d8e960) at src/GafferSceneUI/LightTool.cpp:1689
GafferSceneUI::LightTool::preRender (this=0x7f5093d8e960) at src/GafferSceneUI/LightTool.cpp:1837
```

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5218/commits/61072a1f2f11066af3c9126d894dcc6f3bddf7e8

murraystevenson commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -0,0 +1,1916 @@
+					std::string shaderAttribute = shader->getType() + ":" + shader->getName();
+
+					auto coneParameterName = Metadata::value<StringData>( shaderAttribute, "coneAngleParameter" );
+					assert( coneParameterName );
```
This assert fails/a release build will crash in the following situation:

- Activate the light tool
- Select any light that isn't a spotLight

```
gaffer: src/GafferSceneUI/LightTool.cpp:646: virtual void {anonymous}::SpotLightHandle::update(GafferScene::ScenePathPtr, const PlugPtr&): Assertion `coneParameterName' failed.

__assert_fail () from /lib64/libc.so.6
(anonymous namespace)::SpotLightHandle::update (this=0xbf64fc0, scenePath=..., editScope=...) at src/GafferSceneUI/LightTool.cpp:646
GafferSceneUI::LightTool::updateHandleInspections (this=0xcb77a10) at src/GafferSceneUI/LightTool.cpp:1677
GafferSceneUI::LightTool::preRender (this=0xcb77a10) at src/GafferSceneUI/LightTool.cpp:1837
```

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5218/commits/61072a1f2f11066af3c9126d894dcc6f3bddf7e8

murraystevenson reviewed (changes requested):
Thanks Eric, I went back and did a bit of wider testing again today, rather than just focusing on the behaviour changes introduced in the previous review, and I've run into some situations where LightTool use can lead to a crash. I've commented inline with the repro cases.

ericmehl commented:
Thanks for pointing those out Murray, I had focused so much on the cone vs. penumbra activation I forgot the same code was being run for all the selection, not just lights.

I went through the whole tool and checked that all of the pointers will be valid when they are used. I fixed problems in a few areas including where you pointed out, and it resulted in some decent logical clarifications too.

When you have a minute, have another try at testing it.

danieldresser-ie commented on src/GafferSceneUI/LightTool.cpp:
```diff
@@ -855,11 +862,24 @@ class SpotLightHandle : public LightToolHandle
+			Inspector::ResultPtr contextConeInspection = m_coneAngleInspector->inspect();
+			Inspector::ResultPtr contextPenumbraInspection = m_penumbraAngleInspector ? m_penumbraAngleInspector->inspect() : nullptr;
+
+			if( !contextConeInspection || ( m_handleType == HandleType::Penumbra && !contextPenumbraInspection ) )
```
It feels a bit weird for me that we grab these Inspections to test if they're null, but we never use them for anything ... it makes it unclear why we need to test them.

Are these different from the coneInspection and penumbraInspection values returned from spotLightHandleAngles ... I can't really tell how they differ? If they are the same, then I would find it clearer to call spotLightHandleAngles first, and then test the values we are actually using for null inspections.

ericmehl replied:
They are slightly different from `spotLightHandleAngles()`. That function returns the handle angles and inspections for the scene location where the handles will be attached, the last selected location. These inspections are done in the current context, which for this purpose is set in a loop over the selected locations starting at line 1699. If the inspection is null, it means we can't do anything for this scene location (probably because it's not a spot light) so the handle should be hidden. 

This way we only show the handles if the handle for all of the selection is visible.

I did that in order to keep the selection handling in `LightTool` and not make each individual tool have to keep track of it, which still sounds like a good idea to me but I'm open to suggestions.

Perhaps a better comment is enough to clarify what's being tested?

danieldresser-ie replied:
Oh, now that you've explained it, that does make sense. I think part of the reason it feels weird is that the loop on line 1699 is visiting all the locations in the selection, but the penumbra logic in visible() is just based on the last selected, so each of the visible() calls ends up evaluating the penumbra of the same location, right? Feels a little bit odd to me, but you've thought about this interface way more than I have. I certainly don't see any reason to hold this up any further, but  maybe a comment wouldn't hurt.

ericmehl replied:
> so each of the visible() calls ends up evaluating the penumbra of the same location, right?

Yes, which is not ideal. This is a unique check to spot light penumbras, as least so far, and I'm not sure where would be a better place to put it. Maybe just in the drawing code?

The same caching optimization the drawing code benefits from applies here too, so I don't think it will lead to performance problems. I'm not quite ready to change it without a more thorough audit of the interface overall, so I've added it as a to-do for now.

Maybe after the quad light tool is done I can revisit this overall and have a good picture of what improvements can be made for future maintainability.

For now I've squashed more explanatory comments into https://github.com/GafferHQ/gaffer/pull/5218/commits/49b8945d19d6aecda49d10c3cba482e97b4843cc

danieldresser-ie replied:
Looks fine for now.

danieldresser-ie reviewed (commented):
LGTM, with one comment that I think might affect clarity, but not correctness.

murraystevenson commented:
Thanks Eric! The latest round of updates look good to me for a 1.3 release.

There are a couple of additional niceties that the LightTool should support in order to bring it in line with the other manipulators, but I think we're safe to address these in a future `1.3.x.x` release:

- Handles don't disable when plugs are driven by a non-editable input, such as an Expression, ContextQuery, etc.
- The LightTool is unable to edit animated plugs.

In both cases we currently see `ERROR : EventSignalCombiner : Cannot set value for plug {plug} except during computation.` errors while attempting to edit plugs with animation, expressions, etc as inputs.

ericmehl commented:
Squashed and ready to merge!
